graph/src/ipfs/cache.rs

```diff
@@ -0,0 +1,286 @@
+use std::{
+    path::PathBuf,
+    sync::{Arc, Mutex},
+    time::Duration,
+};
+
+use anyhow::anyhow;
+use async_trait::async_trait;
+use bytes::Bytes;
+use graph_derive::CheapClone;
+use lru_time_cache::LruCache;
+use object_store::{local::LocalFileSystem, path::Path, ObjectStore};
+use redis::{
+    aio::{ConnectionManager, ConnectionManagerConfig},
+    AsyncCommands as _, RedisResult, Value,
+};
+use slog::{debug, info, warn, Logger};
+use tokio::sync::Mutex as AsyncMutex;
+
+use crate::{env::ENV_VARS, prelude::CheapClone};
+
+use super::{
+    ContentPath, IpfsClient, IpfsError, IpfsRequest, IpfsResponse, IpfsResult, RetryPolicy,
+};
+
+struct RedisClient {
+    mgr: AsyncMutex<ConnectionManager>,
+}
+
+impl RedisClient {
+    async fn new(logger: &Logger, path: &str) -> RedisResult<Self> {
+        let env = &ENV_VARS.mappings;
+        let client = redis::Client::open(path)?;
+        let cfg = ConnectionManagerConfig::default()
+            .set_connection_timeout(env.ipfs_timeout)
+            .set_response_timeout(env.ipfs_timeout);
+        info!(logger, "Connecting to Redis for IPFS caching"; "url" => path);
+        // Try to connect once synchronously to check if the server is reachable.
+        let _ = client.get_connection()?;
+        let mgr = AsyncMutex::new(client.get_connection_manager_with_config(cfg).await?);
+        info!(logger, "Connected to Redis for IPFS caching"; "url" => path);
+        Ok(RedisClient { mgr })
+    }
+
+    async fn get(&self, path: &ContentPath) -> IpfsResult<Bytes> {
+        let mut mgr = self.mgr.lock().await;
+
+        let key = Self::key(path);
+        let data: Vec<u8> = mgr
+            .get(&key)
+            .await
+            .map_err(|e| IpfsError::InvalidCacheConfig {
+                source: anyhow!("Failed to get IPFS object {key} from Redis cache: {e}"),
+            })?;
+        Ok(data.into())
+    }
+
+    async fn put(&self, path: &ContentPath, data: &Bytes) -> IpfsResult<()> {
+        let mut mgr = self.mgr.lock().await;
+
+        let key = Self::key(path);
+        mgr.set(&key, data.as_ref())
+            .await
+            .map(|_: Value| ())
+            .map_err(|e| IpfsError::InvalidCacheConfig {
+                source: anyhow!("Failed to put IPFS object {key} in Redis cache: {e}"),
+            })?;
+        Ok(())
+    }
+
+    fn key(path: &ContentPath) -> String {
+        format!("ipfs:{}", path.cid())
```

Comment:
This ensures that we generate the correct key for requests of files within IPFS directories

```suggestion
        format!("ipfs:{path}")
```

Reply:
Aah .. yes! Nice catch!